### PR TITLE
Introduce broker client to service layer

### LINKED ISSUE #19524: Move `brokerClient` handling from REST controllers to service layer
## Description

* Introduce services for user task and job resources, similar to the [ProcessInstanceServices](https://github.com/camunda/camunda/blob/main/service/src/main/java/io/camunda/service/ProcessInstanceServices.java).
* Handle `brokerClient` communication in the services by introducing methods for resource manipulation, e.g. `public CompletableFuture<UserTaskRecord> completeUserTask(final long userTaskKey, final DirectBuffer variables, final String action)`.
* Service methods create and send broker command requests, asynchronously handle the response, map broker errors and rejections to a common `CamundaServiceException` (allowing access to the rejection and error), and return completable futures.
* The REST controllers validate incoming request data, extract the data for the service methods, call the service methods, and handle the response and potential exception asynchronously like they already do, mapping them to respective REST responses.

## Reasoning

* Unified handling of lower-level communication from REST controllers (e.g. to ES/OS and brokers).
* Simplified REST controllers with service access only, nothing more.
* Clear separation of concerns between REST controllers and services.
* Simplified test setup for REST controllers since only service method calls need to be mocked instead of `BrokerClient` communication.

### PR DESCRIPTION
**Pull request by tmetzke**
## Description

* Introduce services for user tasks to handle broker communication.
* Introduce services for jobs to handle broker communication and job activation handlers.
* Use the services from the respective REST controllers instead of handling the broker client there specifically.

## Related issues

closes #19524

### CODE REVIEW

**Comment by tmetzke** [comment]
@romansmirnov, can you look specifically at the authentication and how I used this here, please? I'm uncertain I used this as intended by the services infrastructure. Is this something that can change with every service call?

**Review by romansmirnov — APPROVED** [review]
@tmetzke, I like the changes very much! LGTM! 🚀 (Went through the code, tested it locally, works for me)

File: service/src/main/java/io/camunda/service/JobServices.java
Comment (koevskinikola) [inline service/src/main/java/io/camunda/service/JobServices.java:70]: 💭 I was getting some yellow flags when I saw this `record` since it is another "DTO" class that we'll need to update every time we extend a certain API. And it's not as flexible as the `Broker*Request` classes. But I guess we have the `service` layer now to support us with flexibility.

File: zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java
Comment (koevskinikola) [inline zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java:53]: 🙃  Just some heavy nitpicking: we can move the logger property at the top of the class just to get it out of the way. But this is just a matter of preference and patterns I've seen in the code.

**Review by koevskinikola — APPROVED** [review]
Looks great @tmetzke, I was getting some nice C7 code vibes from the `service` layer 😄 

I didn't see any problems in the code, just some conceptual questions that aren't blocking the merge.

File: service/src/main/java/io/camunda/service/JobServices.java
Comment (tmetzke) [reply service/src/main/java/io/camunda/service/JobServices.java:70]: I see your point, yeah. I was also wondering about it. But then again, this is just a convenience for the service method parameters, ultimately. We could alternatively add all those parameters to the service method signature. That would just simply be less flexible. If you want to extend the record and get that data in from the outside, well, then you have to extend the service to be able to get the data in. You'd have to do this either way, DTO or method signature.

I also thought about handing in the job record instead of the `ActivateJobsRequest` but it didn't have all the fields we need.

File: service/src/main/java/io/camunda/service/search/core/SearchQueryService.java
Comment (tmetzke) [reply service/src/main/java/io/camunda/service/search/core/SearchQueryService.java:24]: Good catch, I also moved back and forth with this in my head and at this point just figured "Well, let's add it, most services will need it anyway, there will be only very few services soon that don't need it, if any". I share your sentiment though and was very close to introducing a new interface as well. But this complicates things, as you mentioned already. It would be a bit cleaner, interface-wise. But I feel like it's over-engineering an issue that will not be relevant in a couple of days/weeks anymore.

File: zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java
Comment (tmetzke) [reply zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java:53]: OK, yeah, I see what you mean. I will take this up on my list and simply adjust this with the next PR for the next job endpoints if you can live with that? 😄

File: service/src/main/java/io/camunda/service/search/core/SearchQueryService.java
Comment (tmetzke) [reply service/src/main/java/io/camunda/service/search/core/SearchQueryService.java:24]: Another aspect to consider here is that the `SearchQueryService` is an abstract class (the `ApiServices` class is, too). We would need 3 different abstract classes as a result:
1. Only search (no broker client but search types) - no broker interaction required (e.g. search over entities we don't have commands for, like flow nodes).
2. Only broker (broker client but no search types) - no search required (e.g. job controller currently).
3. Both broker and search (broker client and search types) - most services (e.g. user task controller).

Not sure if that really helps then or if it's simpler to imply broker client necessity for all services, even if they only do search. It's simple to inject and to provide, although it might not make sense completely for them to have it.

File: service/src/main/java/io/camunda/service/JobServices.java
Comment (koevskinikola) [reply service/src/main/java/io/camunda/service/JobServices.java:70]: Yeah, I think the current solution is the best one. 👍

File: service/src/main/java/io/camunda/service/search/core/SearchQueryService.java
Comment (koevskinikola) [reply service/src/main/java/io/camunda/service/search/core/SearchQueryService.java:24]: I agree, separating it into an interface would be just overengineering due to a naming/conceptual issue.

Maybe another alternative is to rename the `SearchQueryService` class to something that isn't `ES/OS`-specific?

Since this is more of a nitpick 🙃 now as we both agree that the current implementation is good, I'll leave it up to you to decide how to proceed.

Comment (tmetzke) [reply service/src/main/java/io/camunda/service/search/core/SearchQueryService.java:24]: I think it's OK like this. The API services contain the broker client already. The search service is a specialization, it simply inherits the broker client dependency. The job service, for example, is a broker-only service, no search needed there for now. It extends the API services directly. Thus, all API services are automatically broker services, somehow.

**Pull request merged**
